domain/permission/service/service.go

```diff
@@ -0,0 +1,185 @@
+
+	// DeletePermission removes the given user's access to the given target.
+	DeletePermission(ctx context.Context, subject names.UserTag, target names.Tag) error
+
+	// UpdatePermission updates the user's access to the given target to the
+	// given access.
+	UpdatePermission(ctx context.Context, subject names.UserTag, target names.Tag, access permission.Access) error
+
+	// ReadUserAccessForTarget returns the user access for the given user on
+	// the given target.
+	ReadUserAccessForTarget(ctx context.Context, subject names.UserTag, target names.Tag) (permission.UserAccess, error)
+
+	// ReadUserAccessLevelForTarget returns the user access level for the
+	// given user on the given target.
+	ReadUserAccessLevelForTarget(ctx context.Context, subject names.UserTag, target names.Tag) (permission.Access, error)
+
+	// ReadAllUserAccessForUser returns a slice of the user access the given
+	// user has for any access type.
+	ReadAllUserAccessForUser(ctx context.Context, subject names.UserTag) ([]permission.UserAccess, error)
+
+	// ReadAllUserAccessForTarget return a slice of user access for all users
+	// with access to the given target.
+	ReadAllUserAccessForTarget(ctx context.Context, target names.Tag) ([]permission.UserAccess, error)
+
+	// ReadAllAccessTypeForUser return a slice of user access for the user
+	// specified and of the given access type.
+	// E.G. All clouds the user has access to.
+	ReadAllAccessTypeForUser(ctx context.Context, subject names.UserTag, accessType permission.AccessType) ([]permission.UserAccess, error)
+}
+
+// UserAccessSpec defines the attributes that can be set when adding a new
+// user access.
+type UserAccessSpec struct {
+	User       names.UserTag
+	Target     names.Tag
+	Access     permission.Access
+	AccessType permission.AccessType
+}
+
+// Validate validates that the access and access type specified in the
+// spec are values allowed.
+func (u UserAccessSpec) validate() error {
+	if err := validateTarget(u.Target); err != nil {
+		return err
+	}
+	if err := permission.ValidateAccessForAccessType(u.Access, u.AccessType); err != nil {
+		return err
+	}
+	return nil
+}
+
+type Service struct {
+	st State
+}
+
+// NewService returns a new Service for interacting with the underlying
+// permission state.
+func NewService(st State) *Service {
+	return &Service{
+		st: st,
+	}
+}
+
+// CreatePermission gives the user access per the provided spec. An error is
+// returned if the access and access type of the spec are not valid.
+func (s *Service) CreatePermission(ctx context.Context, spec UserAccessSpec) (permission.UserAccess, error) {
+	if err := spec.validate(); err != nil {
+		return permission.UserAccess{}, errors.Trace(err)
+	}
+	userAccess, err := s.st.CreatePermission(ctx, spec)
+	return userAccess, errors.Trace(err)
+}
+
+// DeletePermission removes the given user's access to the given target.
+// An error is returned if the given target's access type does not exist.
```

Comment:
This isn't enforced through policy but a few of the service layers are doing it. Your call if you want to introduce it....

Would you consider making a typed error to represent this error case? At the moment we say "An error is returned for case x". But with this comment the caller has no way to know if that error returned is for that reason or for a completely different reason.

There also does exist a `domain/user/errors.NotFound` error as well if the user doesn't exist as well that you can bubble up the stack if you would like as well.

Same comments as above around Tags. I know @wallyworld feels very strongly about them but I don't mind them as URI's.

Reply:
I will keep this under consideration for the state layer. Some of these will be good in the state layer. The errors here are mostly around is the request good.

Some of the understanding of which types for input and errors will evolve over time as this is used.

Reply:
It also feels like any errors directly from the service layer, are programming errors, not ones you'll hunt down later.